### ISSUE TITLE
Continuations not working well with HTTP/2

### ISSUE THREAD

**Issue by UnsungHero97** [eclipse/jetty.project#766]
I'm submitting a new issue to continue our discussion from #501.

I've set up a public repository with a minimal project setup to reproduce the issue I'm seeing with Continuations over HTTP/2. I basically committed my entire IntelliJ project to try to make it as easy as possible to reproduce.

https://github.com/UnsungHero97/continuations-h2

I've been using Jetty Continuations since Jetty 8, and upgrading to Jetty 9 was smooth; currently running `9.3.10.v20160621`. However, I've started working on enabling HTTP/2, and that is when I started seeing this issue. When the incoming request is HTTP/2, it seems like `continuation.suspend()` is not doing anything because exactly 30 seconds after the request is made, the Continuation is somehow resumed even though I specified a 10-minute timeout. I know 30 seconds is the default idle timeout in some places, but I'm not sure if that's the root cause, or if something else is happening; if I simply just turn off HTTP/2, everything works as expected.

Notes:
  1. I'm running Jetty from inside IntelliJ via the jetty-maven-plugin (see [pom.xml](https://github.com/UnsungHero97/continuations-h2/blob/master/pom.xml)); the `jetty:run-forked` Maven task is required
  2. You'll have to modify [`jetty-ssl-context.xml`](https://github.com/UnsungHero97/continuations-h2/blob/master/src/main/config/jetty/jetty-ssl-context.xml) to update the passwords... `jetty.keystore.password`, `jetty.keymanager.password`, `jetty.truststore.password`
  3. You'll have to recreate the [`keystore`](https://github.com/UnsungHero97/continuations-h2/blob/master/src/main/config/jetty/keystore) file for SSL to work

**Comment by sbordet** [comment]
Cannot reproduce, since everything works as expected.

I don't know what you expect `Continuation.suspend()` to do. It does suspend the request, and if you put a smaller timeout, the request will be resumed.

You have a continuation timeout (10 minutes) that is longer than the TLS connector idle timeout (30 seconds, `jetty.ssl.idleTimeout` in `jetty-ssl.xml`).

Therefore, after 30s Jetty will send a HTTP/2 `RST_STREAM` frame to the client and close the connection. This is expected behavior.

Closing since everything works as expected. If you have further information, please reopen.

**Comment by UnsungHero97** [comment]
In the case when HTTP/2 is disabled, setting the continuation timeout to 10 minutes and suspending the continuation results in keeping the request alive for 10 minutes; the continuation does not resume after 30 seconds. After the 10 minutes are up, the continuation is resumed.

My expectation was that I'd see the same behavior when HTTP/2 is enabled, but that's not what's happening. I'm trying to understand why there's a difference between HTTP/2 enabled or disabled, and I'd like to figure out how to control the timeout on an HTTP/2 request.

**Comment by sbordet** [comment]
All right, the Servlet semantic.

@gregw what do you think ? Should I have a special logic in HTTP/2 that if a stream is suspended, we not only ignore the stream timeout, but we also ignore the connection timeout ?

**Comment by gregw** [comment]
It is a difficult one.   It has been a long time lament of the servlet spec (well JVM really), that a long running servlet (be it async or blocking), is not informed if the connection has closed.   The only way a http1 servlet can find out if a connection is closed is by trying to read or write to it.    Thus we often get the complain that a servlet has spent a long time producing response for a connection that was closed much earlier.

But with http2, because it is a multi-plexed connection, we read and write to it all the time, so we can detect connection failure, plus there a per-stream messages for early abnormal closing.

Thus we do know if a connection has close.... but we now lack the API to tell the servlet that is the case in most cases.  However, if the servlet is async and thye have registered an `AsyncListener`, then we can call the `onError` method to inform the servlet of an early close.   We definitely should do this because it has been such an often requested feature!     @sbordet  is this what we are doing?

I don't thinks so, because looking at the `Servlet3Continuation`, I see that the implementation of onError is empty.    Which is perhaps reasonable, as it will give Continuations the same old behaviour as before (they don't have an onError API).

Ah maybe we are calling onError, but because nothing is done we are then doing an error dispatch, which is a bit silly as we know the connection is closed.      Which is a hassle because Continuations handle a timeout as a resume dispatch.

So maybe we need to add a default `onError` to `ContinuationListener`, whose default behaviour is to complete the continuation with a fake onTimeout?  

So in summary,  we should not ignore the connection timeout, the connection should definitely be able to close for whatever reason while the request is suspended - just like http1.

To me it is just a question of what then should be done with the inError callback within Continuations.


Note we still have to handle request that don't have AsyncListeners...

**Comment by sbordet** [comment]
@gregw we discussed this already in #238, where you were not keen of notifying this condition via `AsyncListener`s, but rather `ServletRequestListener`s.

**Comment by gregw** [comment]
That API has not yet been accepted and there is an alternative version proposed.  Conversation on the thread has stopped and the 4.0 API could be a long way off.

Even with the 4.0 API, we probably would need also notify AsyncContext.onError  as you could have two different libraries working on the same request and each could use a different mechanism.

So we must currently be notifying via `onError` now?  How else are we getting the different continuation behaviour?

**Comment by UnsungHero97** [comment]
any updates?

**Comment by UnsungHero97** [comment]
@sbordet after playing around with this some more, I better understand the current behavior, and I see that `jetty.ssl.idleTimeout` is what is causing the connection to close. What I still don't understand is why the browser retries the request several times after the idle timeout. Is this related to HTTP/2? As you said, Jetty sends an `RST_STREAM` frame to close the connection, so is it expected behavior for a browser to retry the request after this happens?

In the world of HTTP/1.1, I built a service that establishes a connection between 2 clients, A and B, so that Client A can send some data to Client B. Client A waits on Client B to come online, and the wait is capped at a max amount of time, say 10 minutes. Once Client B is online and ready, Client A starts sending data to Client B. Jetty Continuations made this flow easy by allowing me to suspend the request from Client A with a timeout of 10 minutes, and resume the request when Client B is ready. From Client A's perspective, the HTTP request is kept alive while waiting for Client B.

Perhaps there's a different way to do what I've been doing so far in the world of HTTP/2...?

**Comment by UnsungHero97** [comment]
Tangentially, `continuation.setTimeout(0)` does not behave the same way when HTTP/2 is enabled.

**Comment by sbordet** [comment]
@UnsungHero97 there is something that we need to investigate better, we'll keep you posted here.

**Comment by sbordet** [comment]
@UnsungHero97 what is the difference you see in `continuation.setTimeout(0)` ?

**Comment by sbordet** [comment]
There are a number of issues here.

@gregw:

With HTTP/1.1:

1A. Client closes; server cannot detect, `AsyncListener.onError()` not called.
2A. Server idle timeout; ignored because of the logic in `AbstractEndPoint.onIdleExpired()`, `AsyncListener.onError()` not called, connection is kept open forever.

I vaguely remember that this was discussed in the Servlet EG and decided for the current behavior, do I remember well ? 

In HTTP/2:

2A. Client send RST_STREAM; currently ignored because of issue #824, `AsyncListener.onError()` not called.
2B. Client send GOAWAY and closes the connection; server closes the connection, `AsyncListener.onError()` not called because of #824.
2C. Server _stream_ idle timeout; currently ignored because of issue #824, `AsyncListener.onError()` not called.
2D. Server _session_ idle timeout; server sends GOAWAY and closes the connection, `AsyncListener.onError()` not called because of #824.

We already have a difference between HTTP/1.1 and HTTP/2 because in HTTP/2 we do close the connection in case of server session idle timeout.

We can't implement 1A, so if we do implement 2A and 2B we will have a behavior difference.
Implementation of 1B can be changed to call `onError()` and we can implement 2C and 2D to match that and call `onError()` too.

@UnsungHero97:

you are strongly suggested to move away from continuations, since they don't report errors.

I tried your project, and I don't see that with HTTP/2 the request is being resumed after 30s.

Can you provide detailed logs that show this ? Add `slf4j-log4j12` dependency to the Jetty Maven Plugin and specify `-Dlog4j.configuration=file:///...` as a JVM argument, with the location of the log4j configuration file.

**Comment by UnsungHero97** [comment]
Thanks for the detailed overview, @sbordet! I'll get you those logs as soon as I can.

I've been working on transitioning away from Continuations to AsyncContext, but still, the idle timeout is still giving me a hard time. I'll work on minimizing the number of external dependencies to see if the problem is elsewhere, but if I still think it's Jetty, I'll put up another example project to reproduce.

**Comment by UnsungHero97** [comment]
@sbordet Regarding this:

> I tried your project, and I don't see that with HTTP/2 the request is being resumed after 30s.

I misspoke. The request is not resumed. What I meant by "somehow resumed" is that the request somehow fails/closes, which we've now established is due to the idle timeout.

I updated my project with logging. I couldn't figure out how to configure `slf4j-log4j12`, but I think I got `logback` working. Here are the logs from when I make the request to `https://localhost:8443/` with HTTP/2 enabled.

[jetty_logs.txt](https://github.com/eclipse/jetty.project/files/412628/jetty_logs.txt)

**Comment by UnsungHero97** [comment]
I made a new `async-context` branch to try the same thing but with `AsyncContext` instead of Continuations.

https://github.com/UnsungHero97/continuations-h2/tree/async-context

Here are the logs from when I make the request to https://localhost:8443/ with HTTP/2 enabled: 
[jetty_logs_asynccontext.txt](https://github.com/eclipse/jetty.project/files/412653/jetty_logs_asynccontext.txt)

**Comment by sbordet** [comment]
After discussion we are going to align the HTTP/1.1 and HTTP/2 behavior. This is currently implemented in branch `issue_824`, and pending the fix for #833. Can you try that ?

**Comment by UnsungHero97** [comment]
I cloned the Jetty repo, and built the Jetty distribution:

`jetty-distribution/target/jetty-distribution-9.3.12-SNAPSHOT.zip`.

How do I get the `jetty-maven-plugin` to use this snapshot?

**Comment by jmcc0nn3ll** [comment]
If you built it locally then using the 9.3.12-SNAPSHOT version with your jetty-maven-plugin declaration in the build should use the one out of your locale repository.  The jetty-maven-plugin doesn't make use of the distribution at all.  It is probably worth reiterating that you should never use jetty-maven-plugin in a production setting, it is a development tool.

**Comment by UnsungHero97** [comment]
Thanks for your response, @jmcc0nn3ll! Yes, I'm only using `jetty-maven-plugin` for local development, and to make it easy to run Jetty from within IntelliJ.

Cool! I see the snapshot build in `~/.m2/repository/org/eclipse/jetty/jetty-distribution/9.3.12-SNAPSHOT`, and IntelliJ is picking it up. Thanks!

**Comment by UnsungHero97** [comment]
@sbordet I ran the `async-context` branch of my example project against the `issue_824` branch, and I'm still seeing the same issue. Even with `AsyncContext` instead of Continuations, the idle timeout still triggers the reset frame.

Logs: [jetty_logs_asynccontext_issue824.txt](https://github.com/eclipse/jetty.project/files/414766/jetty_logs_asynccontext_issue824.txt)

Note, I changed the various timeout values locally to be smaller for quicker turnaround:

    1. asyncContext.setTimeout(5000);
    2. jetty-ssl.xml: <Set name="idleTimeout"><Property name="jetty.ssl.idleTimeout" deprecated="ssl.timeout" default="1000"/></Set>
    3. jetty-http.xml: <Set name="idleTimeout"><Property name="jetty.http.idleTimeout" deprecated="http.timeout" default="3000"/></Set>

**Comment by sbordet** [comment]
@UnsungHero97 works for me. I changed the idle timeout to 5 seconds, but for the rest is your project untouched.

Request arrives:
```
[STDOUT] 10:26:54.569 [qtp392292416-16] DEBUG org.eclipse.jetty.http2.server.HttpChannelOverHTTP2 - HTTP2 Request #1/3bef6c85, delayed=false:
[STDOUT] GET https://localhost:8443/ HTTP/2.0
...
```
Servlet is called, result WAIT means the request has been suspended:
```
...
[STDOUT] 10:26:54.688 [qtp392292416-26] DEBUG org.eclipse.jetty.server.HttpChannel - ServerHttpChannelOverHTTP2@3869f3af{r=1,c=false,a=ASYNC_WAIT,uri=https://localhost:8443/}#1 handle exit, result WAIT
...
```
First idle timeout expires but it's ignored for both the stream and the session:
```
[STDOUT] 10:26:59.543 [Scheduler-1983747920] DEBUG org.eclipse.jetty.http2.HTTP2Stream - Idle timeout 5000ms expired on HTTP2Stream@3620e2bf#1{sendWindow=6291456,recvWindow=65535,reset=false,REMOTELY_CLOSED}
[STDOUT] 10:26:59.544 [Scheduler-1983747920] DEBUG org.eclipse.jetty.http2.HTTP2Connection - Ignoring idle timeout on HTTP2Stream@3620e2bf#1{sendWindow=6291456,recvWindow=65535,reset=false,REMOTELY_CLOSED}: {}
...
[STDOUT] 10:26:59.572 [Scheduler-1983747920] DEBUG org.eclipse.jetty.io.IdleTimeout - SelectChannelEndPoint@ce04cf9{/0:0:0:0:0:0:0:1:39212<->8443,Open,in,out,FI,-,5001/5000,SslConnection@3d264745}{io=1/1,kio=1,kro=1} idle timeout expired
[STDOUT] 10:26:59.572 [Scheduler-1983747920] DEBUG org.eclipse.jetty.http2.HTTP2Connection - Ignoring idle timeout on HTTP2ServerSession@3bef6c85{l:/0:0:0:0:0:0:0:1:8443 <-> r:/0:0:0:0:0:0:0:1:39212,queueSize=0,sendWindow=15728640,recvWindow=65535,streams=1,NOT_CLOSED}: {}
```
Second idle timeout expires, but it's ignored again:
```
[STDOUT] 10:27:04.545 [Scheduler-1983747920] DEBUG org.eclipse.jetty.http2.HTTP2Stream - Idle timeout 5000ms expired on HTTP2Stream@3620e2bf#1{sendWindow=6291456,recvWindow=65535,reset=false,REMOTELY_CLOSED}
[STDOUT] 10:27:04.545 [Scheduler-1983747920] DEBUG org.eclipse.jetty.http2.HTTP2Connection - Ignoring idle timeout on HTTP2Stream@3620e2bf#1{sendWindow=6291456,recvWindow=65535,reset=false,REMOTELY_CLOSED}: {}
...
[STDOUT] 10:27:04.573 [Scheduler-1983747920] DEBUG org.eclipse.jetty.io.IdleTimeout - SelectChannelEndPoint@ce04cf9{/0:0:0:0:0:0:0:1:39212<->8443,Open,in,out,FI,-,5000/5000,SslConnection@3d264745}{io=1/1,kio=1,kro=1} idle timeout expired
[STDOUT] 10:27:04.574 [Scheduler-1983747920] DEBUG org.eclipse.jetty.http2.HTTP2Connection - Ignoring idle timeout on HTTP2ServerSession@3bef6c85{l:/0:0:0:0:0:0:0:1:8443 <-> r:/0:0:0:0:0:0:0:1:39212,queueSize=0,sendWindow=15728640,recvWindow=65535,streams=1,NOT_CLOSED}: {}
```
Finally the `AsyncContext` timeout fires (it had a 10 seconds timeout):
```
[STDOUT] 10:27:04.688 [Scheduler-1983747920] DEBUG org.eclipse.jetty.server.HttpChannelState - Async timeout HttpChannelState@4bb81078{s=ASYNC_WAIT a=EXPIRING i=false r=!P!U w=false}
[STDOUT] 10:27:04.690 [Scheduler-1983747920] INFO  MinimalServlet - MinimalServlet:wait():onTimeout
```
Response is committed and I see it in the browser.

**Comment by UnsungHero97** [comment]
Awesome. It's great that you got it working!

It seems like you've merged the `issue_824` branch into `9.3.x`. I'm pretty sure I'm on the latest commit on the `9.3.x` branch; `git log` shoes your recent commits, `ea20a5f2ef5fadc6e374809a0a4ffee08fead5a0` and `9f9e9ecaf95ac370b206a627a88fbe64beb69ae7`. Unfortunately, I haven't been able to get it to work locally. Also, I'm fairly certain that IntelliJ and the `jetty-maven-plugin` are picking up the correct `9.3.12-SNAPSHOT` version. I ran `mvn clean install`, and it seems like everything succeeded:

```
...
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 31:50 min
[INFO] Finished at: 2016-08-12T10:06:15-07:00
[INFO] Final Memory: 292M/1821M
[INFO] ------------------------------------------------------------------------
```

The difference between what you have and what I have is that I don't see `Ignoring idle timeout on HTTP2Stream`. Instead...

```
[STDOUT] 11:11:11.911 [Scheduler-345281752] DEBUG org.eclipse.jetty.http2.HTTP2Stream - Idle timeout 5000ms expired on HTTP2Stream@ec5359b#1{sendWindow=6291456,recvWindow=65535,reset=false,REMOTELY_CLOSED}
[STDOUT] 11:11:11.912 [Scheduler-345281752] DEBUG org.eclipse.jetty.http2.HTTP2Session - Sending ResetFrame@3529de8c#1{cancel_stream_error}
[STDOUT] 11:11:11.912 [Scheduler-345281752] DEBUG org.eclipse.jetty.http2.HTTP2Flusher - Appended ResetFrame@3529de8c#1{cancel_stream_error}, frames=1
[STDOUT] 11:11:11.912 [Scheduler-345281752] DEBUG org.eclipse.jetty.http2.HTTP2Flusher - Flushing HTTP2ServerSession@525c4c77{l:/0:0:0:0:0:0:0:1:8443 <-> r:/0:0:0:0:0:0:0:1:57697,queueSize=1,sendWindow=15728640,recvWindow=65535,streams=1,NOT_CLOSED}
[STDOUT] 11:11:11.912 [Scheduler-345281752] DEBUG org.eclipse.jetty.http2.HTTP2Flusher - Gathered for write ResetFrame@3529de8c#1{cancel_stream_error}
[STDOUT] 11:11:11.912 [Scheduler-345281752] DEBUG org.eclipse.jetty.http2.HTTP2Session - Generated ResetFrame@3529de8c#1{cancel_stream_error}
[STDOUT] 11:11:11.912 [Scheduler-345281752] DEBUG org.eclipse.jetty.http2.HTTP2Flusher - Writing 1 buffers (13 bytes) for 1 frames [ResetFrame@3529de8c#1{cancel_stream_error}]
```

I'm not sure what's different about my local environment vs yours. Are you doing `jetty:run-forked`?

**Comment by sbordet** [comment]
@UnsungHero97, yes I merged the branch into `jetty-9.3.x` and yes I use `jetty:run-forked`.

Have you built Jetty locally ?
If not, I think Maven will download the nightly snapshots that may not have the latest changes.

What client are you using, a browser ?

Next thing to do to verify you're running the right code is to debug. Are you hitting the new code ?

**Comment by UnsungHero97** [comment]
Got it! Silly mistake... I only updated `jetty-maven-plugin` to have version `9.3.12-SNAPSHOT`. The rest of `pom.xml` was using `${jettyVersion}`, which was still set to `9.3.10.v20160621`.

Thank you for tackling this, @sbordet! I look forward to the next release!
